drv/lpc55-update-server/src/main.rs

```diff
@@ -226,6 +234,203 @@ impl idl::InOrderUpdateImpl for ServerImpl {
+                let cfpa_bytes: &[u8] = cfpa.as_bytes();
+                let cfpa_bytes: &[u8; BLOCK_SIZE_BYTES] =
+                    cfpa_bytes.try_into().unwrap_lite();
+
+                // Erase and program the scratch page. Note that because the
+                // scratch page is _not_ the authoritative copy, and because the
+                // ROM will check its contents before making it authoritative,
+                // we can fail during this operation without corrupting anything
+                // permanent. Yay!
+                //
+                // Note that the page write machinery uses page numbers. This
+                // should probably change. But, for now, we must divide our word
+                // number by 32.
+                do_raw_page_write(&mut self.flash, 0x9DE0 / 32, &cfpa_bytes)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Reads an arbitrary contiguous set of flash words from flash, indirectly,
+/// using the flash controller interface. This allows access to sections of
+/// flash that are not direct-mapped into our task's memory, saving MPU regions.
+///
+/// `flash_word_number` is (as its name suggests) a _word number,_ a 0-based
+/// index of a 16-byte word within flash where reading should begin.
+///
+/// `output` implies the length to read.
+///
+/// The main way this can fail is by ECC error; currently, if this occurs, no
+/// feedback is provided as to _where_ in the region the error occurred. We may
+/// wish to fix this.
+///
+/// This API produces flash words in the form of `[u32; 4]`, because that's how
+/// the hardware produces them. Elements of the array are in ascending address
+/// order when the flash is viewed as bytes. The easiest way to view the
+/// corresponding block of 16 bytes is using `zerocopy::AsBytes` to reinterpret
+/// the array in place.
+fn indirect_flash_read(
+    flash: &mut drv_lpc55_flash::Flash<'_>,
+    flash_word_number: u32,
+    output: &mut [[u32; 4]],
+) -> Result<(), UpdateError> {
+    use drv_lpc55_flash::ReadError;
+
+    for (wn, dest) in (flash_word_number..).zip(output) {
+        flash.start_read(wn);
+        loop {
+            // Reading is relatively fast; this loop will most likely not sleep,
+            // most of the time.
+            if let Some(result) = flash.poll_read_result() {
+                match result {
+                    Ok(data) => {
+                        *dest = data;
+                        break;
+                    }
+                    Err(ReadError::IllegalOperation) => {
+                        return Err(UpdateError::Unknown);
+                    }
+                    Err(ReadError::Ecc) => {
+                        return Err(UpdateError::EccDoubleErr);
+                    }
+                    Err(ReadError::Fail) => {
+                        return Err(UpdateError::Unknown);
+                    }
+                }
+            }
+
+            // But just in case it needs to:
```

Comment:
You're not clearing the HW interrupt condition on error returns.

Reply:
In the failure cases, lines 408, 411, and 414, the interrupt condition has been polled.
The code starting at 421 shows a desire to clear the lingering HW interrupt condition that has
been left asserted from doing polled IO.
Without looking at the datasheet, it's implied that for this device, merely checking on interrupt status is not enough to clear the HW interrupt condition. Hence the enable and sys_recv_closed calls.

The Result should probably be taken from the match above, then HW interrupt clearing, then return Result.

Reply:
Is line 422 going to hang if there was no interrupt?

Reply:
Can you expand on that? I think the code is okay...

- The status bits in the flash controller are cleared before any operation
- The interrupt enables are cleared on line 429
- There are no early-exit opportunities between when they're enabled on line 421 and cleared on line 429

Reply:
Okay, reading the comments I had missed:

The `poll` functions have no side effects. The interrupt status bits on the device are not cleared. They're only cleared on issuing a new operation, e.g. in this case `start_read`.

This is okay because for the status bits to turn into an actual interrupt, two things need to happen:
- The corresponding interrupt _enable_ bits in the device need to be set (which, when we're going to sleep, we do with the `enable_interrupt_sources` call.
- The hardware interrupt routing in the kernel needs to be enabled (`sys_irq_control`).

When these two conditions are met, an interrupt will be delivered when available and OR'd into the task's notification mask. The task will observe this on the next compatible call to `sys_recv`, which in our case we do immediately. However, those calls have no effect on the interrupt status or enable bits at the device. So, after `sys_recv_closed`, we call `disable_interrupt_sources` on the flash driver to clear the interrupt enables.

This may leave interrupt _status_ bits set (and in practice it absolutely will leave at least one bit set). This is okay, because they don't do anything until the corresponding interrupt source is enabled, and we're going to clear them before doing that next.

Re: your last question -- if the interrupt never comes, the `sys_recv_closed` call will indeed block forever waiting for it.

Reply:
(more careful reading...) Line 402 clears any pending flash controller status flags from the flash controller. The NVIC doesn't get an interrupt until/unless the flash controller DONE bit is set and line 426 and 427 are executed. So, this works out fine.
In the error cases, status bits are set but they are not resulting in NVIC interrupts. The error cases will leave those status bits set, but on the next operation (if the calling layer has not bailed out), those bits are cleared before the next flash I/O operation.

Reply:
So, not a problem.